src/epics/chatEpics.js

```diff
@@ -0,0 +1,212 @@
+import 'rxjs/add/operator/do';
+import 'rxjs/add/operator/retry';
+import 'rxjs/add/operator/pluck';
+import 'rxjs/add/operator/filter';
+import 'rxjs/add/operator/retryWhen';
+import 'rxjs/add/observable/of';
+import 'rxjs/add/observable/from';
+import 'rxjs/add/observable/fromEvent';
+import 'rxjs/add/observable/merge';
+import 'rxjs/add/observable/if';
+
+function createKeyPairObservable({ pincode = '' }) {
+  return Observable.create(function (observer) {
+    const urlHash = document.location.hash + pincode;
+    // 1. Get passphrase
+    const {
+      passphrase,
+      isNewPassphrase
+    } = getPassphrase(urlHash);
+
+    // 3. Get email based on passphrase
+    let email = getEmail(passphrase);
+    // 4. Decrypt to get key pair
+    miniLock.crypto.getKeyPair(passphrase, email, (keyPair) => {
+      miniLock.session.keys = keyPair
+      miniLock.session.keyPairReady = true
+      let mID = miniLock.crypto.getMiniLockID(keyPair.publicKey);
+      // 4. When we have keypair, login:
+      observer.next({ passphrase, keyPair, mID, isNewRoom: isNewPassphrase });
+      observer.complete();
+    });
+  });
+}
+
+function createDecryptMessageObservable({ message, mID, secretKey }) {
+  return Observable.create(function (observer) {
+    miniLock.crypto.decryptFile(message, mID, secretKey,
+      function (fileBlob, saveName, senderID) {
+        const reader = new FileReader();
+        reader.addEventListener("loadend", () => {
+          const authToken = reader.result;
+          observer.next(authToken);
+          observer.complete();
+        });
+
+        reader.readAsText(fileBlob);
+      })
+  })
+}
+
+function getAuthRequestSettings({ mID }) {
+  const settings = {
+    url: authUrl,
+    responseType: 'blob',
+    headers: {
+      'X-Minilock-Id': mID
+    },
+    method: 'GET'
+  }
+  return settings;
+}
+
+const initConnectionEpic = (action$) =>
+  action$.ofType(CHAT_INIT_CONNECTION)
+    .mergeMap(createKeyPairObservable)
+    .mergeMap(({ keyPair, mID, isNewRoom, passphrase }) =>
+      Observable.merge(
+        Observable.of(alertSuccess('New room created.'))
+          .filter(() => isNewRoom)
+          .do(() => document.location.hash = '#' + passphrase),
+
+        ajax(getAuthRequestSettings({ mID }))
+          .map(ajaxResult => ({
+            message: ajaxResult.response,
+            mID,
+            secretKey: keyPair.secretKey
+          }))
+          .mergeMap(createDecryptMessageObservable)
+          .mergeMap(authToken =>
+            chatHandler.initConnection({
+              authToken,
+              secretKey: keyPair.secretKey,
+              mID
+            })
+          )
+          .mergeMap(() =>
+            Observable.merge(
+              chatHandler.getMessageSubject()
+                .map(addMessage),
+
+              chatHandler.getUserStatusSubject()
+                .map(setUserStatus),
+
+              Observable.from([connectionInitiated(), alertSuccess('Connected.')])
```

Comment:
@Spetastian Unfortunately `alertSuccess('Connected.')` shows up like ~50ms after `Observable.of(alertSuccess('New room created.'))`, which results in `New room created` not being seen.

How about removing the `Connected.` message until we change how Alerts work so that multiple can be showing at once?

Reply:
Good idea. Also this Rx flow might need som refactoring as well, but it works really fine for now.